File: compat/suites/python-sdk/groups/kinesis.py
"""
groups/kinesis.py — Kinesis compatibility test implementations.
"""

from __future__ import annotations
import base64
import time
from lib.harness import TestContext
from lib.clients import make_clients


def _kin(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region).kinesis


def _wait_stream_active(kin, stream_name: str, max_wait: int = 10) -> None:
    deadline = time.time() + max_wait
    while time.time() < deadline:
        resp = kin.describe_stream_summary(StreamName=stream_name)
        status = resp["StreamDescriptionSummary"]["StreamStatus"]
        if status == "ACTIVE":
            return
        time.sleep(0.2)
    raise TimeoutError(f"Stream {stream_name!r} did not become ACTIVE within {max_wait}s")


# ── kinesis-streams ───────────────────────────────────────────────────────────

def setup_kinesis_streams(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = f"oc-{ctx.run_id}-stream"
    kin.create_stream(StreamName=name, ShardCount=1)
    _wait_stream_active(kin, name)
    ctx["kinesis_stream"] = name


def teardown_kinesis_streams(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx.get("kinesis_stream")
    if not name:
        return
    try:
        kin.delete_stream(StreamName=name, EnforceConsumerDeletion=True)
    except Exception:
        pass


def CreateStream(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = f"oc-{ctx.run_id}-create"
    kin.create_stream(StreamName=name, ShardCount=1)
    _wait_stream_active(kin, name)
    ctx["kinesis_create_stream"] = name
    kin.delete_stream(StreamName=name, EnforceConsumerDeletion=True)


def DescribeStream(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_stream"]
    resp = kin.describe_stream(StreamName=name)
    desc = resp["StreamDescription"]
    if desc["StreamName"] != name:
        raise AssertionError(f"DescribeStream: name mismatch {desc['StreamName']!r}")
    if not desc.get("Shards"):
        raise AssertionError("DescribeStream: no shards returned")


def DescribeStreamSummary(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_stream"]
    resp = kin.describe_stream_summary(StreamName=name)
    summary = resp["StreamDescriptionSummary"]
    if summary["StreamName"] != name:
        raise AssertionError(f"DescribeStreamSummary: name mismatch {summary['StreamName']!r}")
    if summary["OpenShardCount"] < 1:
        raise AssertionError(f"DescribeStreamSummary: expected ≥1 shard, got {summary['OpenShardCount']}")


def ListStreams(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_stream"]
    resp = kin.list_streams()
    if name not in resp.get("StreamNames", []):
        raise AssertionError(f"ListStreams: {name!r} not found in {resp.get('StreamNames')}")


def AddTagsToStream(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_stream"]
    kin.add_tags_to_stream(StreamName=name, Tags={"env": "compat", "suite": "python-sdk"})
    resp = kin.list_tags_for_stream(StreamName=name)
    tags = {t["Key"]: t["Value"] for t in resp.get("Tags", [])}
    if tags.get("env") != "compat":
        raise AssertionError(f"AddTagsToStream: env=compat tag not found after add; got {tags}")


def ListTagsForStream(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_stream"]
    resp = kin.list_tags_for_stream(StreamName=name)
    tags = {t["Key"]: t["Value"] for t in resp.get("Tags", [])}
    if tags.get("env") != "compat":
        raise AssertionError(f"ListTagsForStream: expected env=compat, got {tags}")


def DeleteStream(ctx: TestContext) -> None:
    # Already tested via setup lifecycle; test a separate transient stream
    kin = _kin(ctx)
    name = f"oc-{ctx.run_id}-del"
    kin.create_stream(StreamName=name, ShardCount=1)
    _wait_stream_active(kin, name)
    kin.delete_stream(StreamName=name, EnforceConsumerDeletion=True)
    resp = kin.list_streams()
    if name in resp.get("StreamNames", []):
        raise AssertionError(f"DeleteStream: {name!r} still listed after deletion")


# ── kinesis-records ───────────────────────────────────────────────────────────

def setup_kinesis_records(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = f"oc-{ctx.run_id}-rec"
    kin.create_stream(StreamName=name, ShardCount=1)
    _wait_stream_active(kin, name)
    ctx["kinesis_rec_stream"] = name


def teardown_kinesis_records(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx.get("kinesis_rec_stream")
    if not name:
        return
    try:
        kin.delete_stream(StreamName=name, EnforceConsumerDeletion=True)
    except Exception:
        pass


def PutRecord(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_rec_stream"]
    resp = kin.put_record(
        StreamName=name,
        Data=b"hello-record",
        PartitionKey="pk-1",
    )
    if not resp.get("ShardId"):
        raise AssertionError(f"PutRecord: missing ShardId in {resp}")
    ctx["kinesis_shard_id"] = resp["ShardId"]


def PutRecords(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_rec_stream"]
    records = [
        {"Data": f"record-{i}".encode(), "PartitionKey": f"pk-{i}"}
        for i in range(5)
    ]
    resp = kin.put_records(StreamName=name, Records=records)
    if resp.get("FailedRecordCount", 0) > 0:
        raise AssertionError(f"PutRecords: {resp['FailedRecordCount']} failed records")


def GetShardIterator(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_rec_stream"]
    shard_id = ctx.get("kinesis_shard_id") or "shardId-000000000000"
    resp = kin.get_shard_iterator(
        StreamName=name,
        ShardId=shard_id,
        ShardIteratorType="TRIM_HORIZON",
    )
    it = resp.get("ShardIterator")
    if not it:
        raise AssertionError("GetShardIterator: missing ShardIterator")
    ctx["kinesis_shard_iter"] = it


def GetRecords(ctx: TestContext) -> None:
    kin = _kin(ctx)
    it = ctx.get("kinesis_shard_iter")
    if not it:
        raise AssertionError("GetRecords: no shard iterator in context (run GetShardIterator first)")
    resp = kin.get_records(ShardIterator=it, Limit=100)
    records = resp.get("Records", [])
    if not records:
        raise AssertionError("GetRecords: no records returned (expected ≥1 from PutRecord/PutRecords)")
    # Verify data is readable
    _ = base64.b64decode(records[0]["Data"]) if isinstance(records[0]["Data"], str) else records[0]["Data"]


# ── kinesis-shards ────────────────────────────────────────────────────────────

def setup_kinesis_shards(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = f"oc-{ctx.run_id}-shard"
    kin.create_stream(StreamName=name, ShardCount=2)
    _wait_stream_active(kin, name)
    ctx["kinesis_shard_stream"] = name


def teardown_kinesis_shards(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx.get("kinesis_shard_stream")
    if not name:
        return
    try:
        kin.delete_stream(StreamName=name, EnforceConsumerDeletion=True)
    except Exception:
        pass


def ListShards(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_shard_stream"]
    resp = kin.list_shards(StreamName=name)
    shards = resp.get("Shards", [])
    if len(shards) < 1:
        raise AssertionError(f"ListShards: expected ≥1 shard, got {len(shards)}")


def SplitShard(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_shard_stream"]
    # Get first shard
    resp = kin.list_shards(StreamName=name)
    shards = resp.get("Shards", [])
    if not shards:
        raise AssertionError("SplitShard: no shards to split")
    shard = shards[0]
    # Compute midpoint of the hash range
    start = int(shard["HashKeyRange"]["StartingHashKey"])
    end = int(shard["HashKeyRange"]["EndingHashKey"])
    mid = str((start + end) // 2)
    kin.split_shard(StreamName=name, ShardToSplit=shard["ShardId"], NewStartingHashKey=mid)
    _wait_stream_active(kin, name)
    resp2 = kin.list_shards(StreamName=name)
    open_shards = [s for s in resp2.get("Shards", []) if "EndingSequenceNumber" not in s.get("SequenceNumberRange", {})]
    assert len(open_shards) >= 2, f"SplitShard: expected >=2 open shards, got {len(open_shards)}"


def MergeShards(ctx: TestContext) -> None:
    kin = _kin(ctx)
    name = ctx["kinesis_shard_stream"]
    resp = kin.list_shards(StreamName=name)
    open_shards = [
        s for s in resp.get("Shards", [])
        if "EndingSequenceNumber" not in s.get("SequenceNumberRange", {})
    ]
    if len(open_shards) < 2:
        raise AssertionError(f"MergeShards: need >=2 open shards, got {len(open_shards)}")
    kin.merge_shards(
        StreamName=name,
        ShardToMerge=open_shards[0]["ShardId"],
        AdjacentShardToMerge=open_shards[1]["ShardId"],
    )
    _wait_stream_active(kin, name)
    resp2 = kin.list_shards(StreamName=name)
    after_open = [
        s for s in resp2.get("Shards", [])
        if "EndingSequenceNumber" not in s.get("SequenceNumberRange", {})
    ]
    assert len(after_open) < len(open_shards), (
        f"MergeShards: expected fewer open shards, got {len(after_open)}"
    )


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateStream": CreateStream,
    "DescribeStream": DescribeStream,
    "DescribeStreamSummary": DescribeStreamSummary,
    "ListStreams": ListStreams,
    "AddTagsToStream": AddTagsToStream,
    "ListTagsForStream": ListTagsForStream,
    "DeleteStream": DeleteStream,
    "PutRecord": PutRecord,
    "PutRecords": PutRecords,
    "GetShardIterator": GetShardIterator,
    "GetRecords": GetRecords,
    "ListShards": ListShards,
    "SplitShard": SplitShard,
    "MergeShards": MergeShards,
}

SETUP = {
    "kinesis-streams": setup_kinesis_streams,
    "kinesis-records": setup_kinesis_records,
    "kinesis-shards": setup_kinesis_shards,
}

TEARDOWN = {
    "kinesis-streams": teardown_kinesis_streams,
    "kinesis-records": teardown_kinesis_records,
    "kinesis-shards": teardown_kinesis_shards,
}

File: compat/suites/go-sdk/internal/groups/kinesis.go
package groups

import (
	"context"
	"fmt"
	"math/big"
	"time"

	"github.com/Neaox/overcast-compat-go-sdk/internal/clients"
	"github.com/Neaox/overcast-compat-go-sdk/internal/harness"
	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/kinesis"
	"github.com/aws/aws-sdk-go-v2/service/kinesis/types"
)

func Kinesis(c *clients.Clients) ServiceGroup {
	g := &kinesisGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"CreateStream":          g.CreateStream,
			"DescribeStream":        g.DescribeStream,
			"DescribeStreamSummary": g.DescribeStreamSummary,
			"ListStreams":           g.ListStreams,
			"AddTagsToStream":       g.AddTagsToStream,
			"ListTagsForStream":     g.ListTagsForStream,
			"DeleteStream":          g.DeleteStream,
			"PutRecord":             g.PutRecord,
			"PutRecords":            g.PutRecords,
			"GetShardIterator":      g.GetShardIterator,
			"GetRecords":            g.GetRecords,
			"ListShards":            g.ListShards,
			"SplitShard":            g.SplitShard,
			"MergeShards":           g.MergeShards,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"kinesis-streams": g.setupStreams,
			"kinesis-records": g.setupRecords,
			"kinesis-shards":  g.setupShards,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"kinesis-streams": g.teardownStreams,
			"kinesis-records": g.teardownRecords,
			"kinesis-shards":  g.teardownShards,
		},
	}
}

type kinesisGroup struct{ c *clients.Clients }

func (g *kinesisGroup) cl() *kinesis.Client { return g.c.Kinesis() }

func (g *kinesisGroup) waitStreamActive(ctx context.Context, streamName string) error {
	deadline := time.Now().Add(30 * time.Second)
	for time.Now().Before(deadline) {
		resp, err := g.cl().DescribeStreamSummary(ctx, &kinesis.DescribeStreamSummaryInput{
			StreamName: aws.String(streamName),
		})
		if err != nil {
			return err
		}
		if resp.StreamDescriptionSummary.StreamStatus == types.StreamStatusActive {
			return nil
		}
		time.Sleep(500 * time.Millisecond)
	}
	return fmt.Errorf("stream %q did not become ACTIVE within timeout", streamName)
}

// ── kinesis-streams ───────────────────────────────────────────────────────────

func (g *kinesisGroup) setupStreams(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("oc-stream-%s", t.RunID)
	if _, err := g.cl().CreateStream(ctx, &kinesis.CreateStreamInput{
		StreamName: aws.String(name),
		ShardCount: aws.Int32(1),
	}); err != nil {
		return err
	}
	if err := g.waitStreamActive(ctx, name); err != nil {
		return err
	}
	t.Set("kinesis_stream", name)
	return nil
}

func (g *kinesisGroup) teardownStreams(ctx context.Context, t *harness.TestContext) error {
	if name := t.GetString("kinesis_stream"); name != "" {
		g.cl().DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(name)}) //nolint:errcheck
	}
	return nil
}

func (g *kinesisGroup) CreateStream(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("oc-cs-%s", t.RunID)
	_, err := g.cl().CreateStream(ctx, &kinesis.CreateStreamInput{
		StreamName: aws.String(name),
		ShardCount: aws.Int32(1),
	})
	if err != nil {
		return err
	}
	g.waitStreamActive(ctx, name) //nolint:errcheck
	// Verify stream appears in ListStreams
	list, lErr := g.cl().ListStreams(ctx, &kinesis.ListStreamsInput{})
	if lErr != nil {
		g.cl().DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(name)}) //nolint:errcheck
		return fmt.Errorf("CreateStream: ListStreams verify failed: %w", lErr)
	}
	found := false
	for _, sn := range list.StreamNames {
		if sn == name {
			found = true
			break
		}
	}
	g.cl().DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(name)}) //nolint:errcheck
	if !found {
		return fmt.Errorf("CreateStream: stream %q not found in ListStreams", name)
	}
	return nil
}

func (g *kinesisGroup) DescribeStream(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().DescribeStream(ctx, &kinesis.DescribeStreamInput{
		StreamName: aws.String(t.GetString("kinesis_stream")),
	})
	if err != nil {
		return err
	}
	if resp.StreamDescription == nil {
		return fmt.Errorf("DescribeStream: nil description")
	}
	return nil
}

func (g *kinesisGroup) DescribeStreamSummary(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().DescribeStreamSummary(ctx, &kinesis.DescribeStreamSummaryInput{
		StreamName: aws.String(t.GetString("kinesis_stream")),
	})
	if err != nil {
		return err
	}
	if resp.StreamDescriptionSummary == nil || aws.ToString(resp.StreamDescriptionSummary.StreamName) == "" {
		return fmt.Errorf("DescribeStreamSummary: missing stream name in response")
	}
	return nil
}

func (g *kinesisGroup) ListStreams(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().ListStreams(ctx, &kinesis.ListStreamsInput{})
	if err != nil {
		return err
	}
	found := false
	for _, s := range resp.StreamNames {
		if s == t.GetString("kinesis_stream") {
			found = true
			break
		}
	}
	if !found {
		return fmt.Errorf("ListStreams: stream %q not found", t.GetString("kinesis_stream"))
	}
	return nil
}

func (g *kinesisGroup) AddTagsToStream(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().AddTagsToStream(ctx, &kinesis.AddTagsToStreamInput{
		StreamName: aws.String(t.GetString("kinesis_stream")),
		Tags:       map[string]string{"env": "test"},
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().ListTagsForStream(ctx, &kinesis.ListTagsForStreamInput{
		StreamName: aws.String(t.GetString("kinesis_stream")),
	})
	if err != nil {
		return fmt.Errorf("AddTagsToStream: ListTagsForStream verify failed: %w", err)
	}
	for _, tag := range resp.Tags {
		if aws.ToString(tag.Key) == "env" && aws.ToString(tag.Value) == "test" {
			return nil
		}
	}
	return fmt.Errorf("AddTagsToStream: env=test tag not found")
}

func (g *kinesisGroup) ListTagsForStream(ctx context.Context, t *harness.TestContext) error {
	resp, err := g.cl().ListTagsForStream(ctx, &kinesis.ListTagsForStreamInput{
		StreamName: aws.String(t.GetString("kinesis_stream")),
	})
	if err != nil {
		return err
	}
	if len(resp.Tags) == 0 {
		return fmt.Errorf("ListTagsForStream: expected ≥1 tag")
	}
	return nil
}

func (g *kinesisGroup) RemoveTagsFromStream(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().RemoveTagsFromStream(ctx, &kinesis.RemoveTagsFromStreamInput{
		StreamName: aws.String(t.GetString("kinesis_stream")),
		TagKeys:    []string{"env"},
	})
	if err != nil {
		return err
	}
	resp, err := g.cl().ListTagsForStream(ctx, &kinesis.ListTagsForStreamInput{
		StreamName: aws.String(t.GetString("kinesis_stream")),
	})
	if err != nil {
		return fmt.Errorf("RemoveTagsFromStream: ListTagsForStream verify failed: %w", err)
	}
	for _, tag := range resp.Tags {
		if aws.ToString(tag.Key) == "env" {
			return fmt.Errorf("RemoveTagsFromStream: env tag still present")
		}
	}
	return nil
}

func (g *kinesisGroup) DeleteStream(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("oc-ds-%s", t.RunID)
	g.cl().CreateStream(ctx, &kinesis.CreateStreamInput{StreamName: aws.String(name), ShardCount: aws.Int32(1)}) //nolint:errcheck
	g.waitStreamActive(ctx, name)                                                                                //nolint:errcheck
	_, err := g.cl().DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(name)})
	if err != nil {
		return err
	}
	list, lErr := g.cl().ListStreams(ctx, &kinesis.ListStreamsInput{})
	if lErr != nil {
		return nil
	}
	for _, sn := range list.StreamNames {
		if sn == name {
			return fmt.Errorf("DeleteStream: stream %q still present", name)
		}
	}
	return nil
}

// ── kinesis-records ───────────────────────────────────────────────────────────

func (g *kinesisGroup) setupRecords(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("oc-rec-%s", t.RunID)
	if _, err := g.cl().CreateStream(ctx, &kinesis.CreateStreamInput{
		StreamName: aws.String(name), ShardCount: aws.Int32(1),
	}); err != nil {
		return err
	}
	if err := g.waitStreamActive(ctx, name); err != nil {
		return err
	}
	t.Set("kinesis_rec_stream", name)
	return nil
}

func (g *kinesisGroup) teardownRecords(ctx context.Context, t *harness.TestContext) error {
	if name := t.GetString("kinesis_rec_stream"); name != "" {
		g.cl().DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(name)}) //nolint:errcheck
	}
	return nil
}

func (g *kinesisGroup) PutRecord(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_rec_stream")
	resp, err := g.cl().PutRecord(ctx, &kinesis.PutRecordInput{
		StreamName:   aws.String(stream),
		Data:         []byte("record-data"),
		PartitionKey: aws.String("pk1"),
	})
	if err != nil {
		return err
	}
	if aws.ToString(resp.ShardId) == "" {
		return fmt.Errorf("PutRecord: missing ShardId in response")
	}
	return nil
}

func (g *kinesisGroup) PutRecords(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_rec_stream")
	resp, err := g.cl().PutRecords(ctx, &kinesis.PutRecordsInput{
		StreamName: aws.String(stream),
		Records: []types.PutRecordsRequestEntry{
			{Data: []byte("r1"), PartitionKey: aws.String("pk1")},
			{Data: []byte("r2"), PartitionKey: aws.String("pk2")},
		},
	})
	if err != nil {
		return err
	}
	if aws.ToInt32(resp.FailedRecordCount) > 0 {
		return fmt.Errorf("PutRecords: %d records failed", aws.ToInt32(resp.FailedRecordCount))
	}
	return nil
}

func (g *kinesisGroup) GetShardIterator(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_rec_stream")
	// Get first shard ID
	desc, err := g.cl().DescribeStream(ctx, &kinesis.DescribeStreamInput{StreamName: aws.String(stream)})
	if err != nil {
		return err
	}
	if len(desc.StreamDescription.Shards) == 0 {
		return fmt.Errorf("GetShardIterator: no shards")
	}
	shardID := aws.ToString(desc.StreamDescription.Shards[0].ShardId)
	resp, err := g.cl().GetShardIterator(ctx, &kinesis.GetShardIteratorInput{
		StreamName:        aws.String(stream),
		ShardId:           aws.String(shardID),
		ShardIteratorType: types.ShardIteratorTypeTrimHorizon,
	})
	if err != nil {
		return err
	}
	t.Set("kinesis_shard_iter", aws.ToString(resp.ShardIterator))
	return nil
}

func (g *kinesisGroup) GetRecords(ctx context.Context, t *harness.TestContext) error {
	iter := t.GetString("kinesis_shard_iter")
	if iter == "" {
		return nil
	}
	resp, err := g.cl().GetRecords(ctx, &kinesis.GetRecordsInput{
		ShardIterator: aws.String(iter),
		Limit:         aws.Int32(10),
	})
	if err != nil {
		return err
	}
	if len(resp.Records) == 0 {
		return fmt.Errorf("GetRecords: expected ≥1 record")
	}
	return nil
}

// ── kinesis-shards ────────────────────────────────────────────────────────────

func (g *kinesisGroup) setupShards(ctx context.Context, t *harness.TestContext) error {
	name := fmt.Sprintf("oc-shd-%s", t.RunID)
	if _, err := g.cl().CreateStream(ctx, &kinesis.CreateStreamInput{
		StreamName: aws.String(name), ShardCount: aws.Int32(2),
	}); err != nil {
		return err
	}
	if err := g.waitStreamActive(ctx, name); err != nil {
		return err
	}
	t.Set("kinesis_shd_stream", name)
	return nil
}

func (g *kinesisGroup) teardownShards(ctx context.Context, t *harness.TestContext) error {
	if name := t.GetString("kinesis_shd_stream"); name != "" {
		g.cl().DeleteStream(ctx, &kinesis.DeleteStreamInput{StreamName: aws.String(name)}) //nolint:errcheck
	}
	return nil
}

func (g *kinesisGroup) ListShards(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_shd_stream")
	resp, err := g.cl().ListShards(ctx, &kinesis.ListShardsInput{StreamName: aws.String(stream)})
	if err != nil {
		return err
	}
	if len(resp.Shards) < 2 {
		return fmt.Errorf("ListShards: expected ≥2 shards, got %d", len(resp.Shards))
	}
	return nil
}

func (g *kinesisGroup) SplitShard(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_shd_stream")
	desc, err := g.cl().DescribeStream(ctx, &kinesis.DescribeStreamInput{StreamName: aws.String(stream)})
	if err != nil {
		return err
	}
	if len(desc.StreamDescription.Shards) == 0 {
		return fmt.Errorf("SplitShard: no shards")
	}
	shard := desc.StreamDescription.Shards[0]
	startKey := aws.ToString(shard.HashKeyRange.StartingHashKey)
	endKey := aws.ToString(shard.HashKeyRange.EndingHashKey)

	// compute midpoint as string
	midKey := hashMidpoint(startKey, endKey)
	_, err = g.cl().SplitShard(ctx, &kinesis.SplitShardInput{
		StreamName:         aws.String(stream),
		ShardToSplit:       shard.ShardId,
		NewStartingHashKey: aws.String(midKey),
	})
	if err != nil {
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	return g.waitStreamActive(ctx, stream)
}

func (g *kinesisGroup) MergeShards(ctx context.Context, t *harness.TestContext) error {
	stream := t.GetString("kinesis_shd_stream")
	desc, err := g.cl().DescribeStream(ctx, &kinesis.DescribeStreamInput{StreamName: aws.String(stream)})
	if err != nil {
		return err
	}
	shards := desc.StreamDescription.Shards
	// find two adjacent open shards
	openShards := make([]types.Shard, 0, len(shards))
	for _, s := range shards {
		if s.SequenceNumberRange.EndingSequenceNumber == nil {
			openShards = append(openShards, s)
		}
	}
	if len(openShards) < 2 {
		return nil // not enough open shards to merge, skip
	}
	_, err = g.cl().MergeShards(ctx, &kinesis.MergeShardsInput{
		StreamName:           aws.String(stream),
		ShardToMerge:         openShards[0].ShardId,
		AdjacentShardToMerge: openShards[1].ShardId,
	})
	if err != nil {
		if harness.IsUnimplemented(err) {
			return nil
		}
		return err
	}
	return g.waitStreamActive(ctx, stream)
}

// hashMidpoint computes the decimal midpoint between two decimal 128-bit hash key strings.
func hashMidpoint(start, end string) string {
	s := new(big.Int)
	e := new(big.Int)
	two := big.NewInt(2)
	s.SetString(start, 10)
	e.SetString(end, 10)
	mid := new(big.Int).Add(s, e)
	mid.Div(mid, two)
	return mid.String()
}
